### PR TITLE
Add caching, skipping, and user-specified background colors to the OpenGL renderer

### PR DESCRIPTION
**Pull request by TRoboto**
## Motivation

Config arguments for `from_animation_number` and `upto_animation_number` don't work with OpenGL. The OpenGL play logic doesn't support caching nor manim output format. Also, the OpenGL window background color is hard-coded. 

## Overview / Explanation for Changes

OpenGL play logic has been improved to support caching and skipping with `-n` argument ( it is now similar to Cairo play logic). A random bug was fixed in OpenGLSurface and OpenGL background color can now be changed via `background_color` argument.

## Testing Status

I tested all changes locally, and everything seems fine.
## Further Comments
I think there should be a parent class for all renderers since there are a few shared functions.

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have chosen a descriptive PR title (see top of PR template for examples)

## Reviewer Checklist
- [x] Newly added functions/classes are either private or have a docstring
- [x] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [x] The PR title is descriptive enough

### CODE REVIEW

**Comment by eulertour** [comment]
These changes should be spit up to prevent problems with any one affecting the others. Additionally, the uv_func parameter should just be accepted by OpenGLSurface rather than making a subclass for it.

**Comment by TRoboto** [comment]
I totally agree. I reverted the OpenGLParametricSurface class and I will make a separate PR later taking your feedback into consideration. Now, this PR only improves the OpenGL play logic with a minor change to the window background color variable.

File: manim/renderer/opengl_renderer.py [hunk 1]
```diff
@@ -353,25 +357,28 @@ def init_scene(self, scene):
...
+        """
+        if config["from_animation_number"]:
+            if self.num_plays < config["from_animation_number"]:
+                self.skip_animations = True
+        if config["upto_animation_number"]:
+            if self.num_plays > config["upto_animation_number"]:
+                self.skip_animations = True
+                raise EndSceneEarlyException()
+
+    @pass_scene_reference
```
Comment (eulertour) [inline manim/renderer/opengl_renderer.py:376]: I think we found that pass_scene_reference isn't needed here.

**Review by eulertour — CHANGES_REQUESTED** [review]
I ran into an error running `poetry run manim example_scenes/opengl.py SurfaceExample --use_opengl_renderer -ql -w -n1`:

```
╭──────────────────────────────────────────────────────────────────────────────────────╮
│ File "/home/devneal/github/manim-2/manim/__main__.py", line 84, in main              │
│     81                 try:                                                          │
│     82                     renderer = OpenGLRenderer()                               │
│     83                     scene = SceneClass(renderer)                              │
│  ❱  84                     scene.render()                                            │
│     85                 except Exception:                                             │
│     86                     console.print_exception()                                 │
│     87         elif config["use_webgl_renderer"]:                                    │
│ File "/home/devneal/github/manim-2/manim/scene/scene.py", line 170, in render        │
│    167         """                                                                   │
│    168         self.setup()                                                          │
│    169         try:                                                                  │
│  ❱ 170             self.construct()                                                  │
│    171         except EndSceneEarlyException:                                        │
│    172             pass                                                              │
│    173         self.tear_down()                                                      │
│ File "/home/devneal/github/manim-2/example_scenes/opengl.py", line 99, in construct  │
│     96                                                                               │
│     97         surface = surfaces[0]                                                 │
│     98                                                                               │
│  ❱  99         self.play(                                                            │
│    100             FadeIn(surface),                                                  │
│    101             ShowCreation(surface.mesh, lag_ratio=0.01, run_time=3),           │
│    102         )                                                                     │
│ File "/home/devneal/github/manim-2/manim/scene/scene.py", line 782, in play          │
│    779             return np.max([animation.run_time for animation in animations])   │
│    780                                                                               │
│    781     def play(self, *args, **kwargs):                                          │
│  ❱ 782         self.renderer.play(self, *args, **kwargs)                             │
│    783                                                                               │
│    784     def wait(self, duration=DEFAULT_WAIT_TIME, stop_condition=None):          │
│    785         self.play(Wait(duration=duration, stop_condition=stop_condition))     │
│ File "/home/devneal/github/manim-2/manim/renderer/cairo_renderer.py", line 13, in    │
│wrapper                                                                               │
│     10                                                                               │
│     11 def pass_scene_reference(func):                                               │
│     12     def wrapper(self, scene, *args, **kwargs):                                │
│  ❱  13         func(self, scene, *args, **kwargs)                                    │
│     14                                                                               │
│     15     return wrapper                                                            │
│     16                                                                               │
│ File "/home/devneal/github/manim-2/manim/utils/caching.py", line 27, in wrapper      │
│    24         scene.add_mobjects_from_animations(animations)                         │
│    25         if self.skip_animations:                                               │
│    26             logger.debug(f"Skipping animation {self.num_plays}")               │
│  ❱ 27             func(self, scene, *args, **kwargs)                                 │
│    28             # If the animation is skipped, we mark its hash as None.           │
│    29             # When sceneFileWriter will start combining partial movie files, it│
│    30             self.animations_hashes.append(None)                                │
│ File "/home/devneal/github/manim-2/manim/renderer/cairo_renderer.py", line 42, in    │
│wrapper                                                                               │
│     39     def wrapper(self, scene, *args, **kwargs):                                │
│     40         self.animation_start_time = time.time()                               │
│     41         self.file_writer.begin_animation(not self.skip_animations)            │
│  ❱  42         func(self, scene, *args, **kwargs)                                    │
│     43         self.file_writer.end_animation(not self.skip_animations)              │
│     44         self.num_plays += 1                                                   │
│     45                                                                               │
│ File "/home/devneal/github/manim-2/manim/renderer/opengl_renderer.py", line 381, in  │
│play                                                                                  │
│    378     @handle_play_like_call                                                    │
│    379     def play(self, scene, *args, **kwargs):                                   │
│    380         if scene.compile_animation_data(*args, **kwargs):                     │
│  ❱ 381             scene.play_internal()                                             │
│    382                                                                               │
│    383     def render(self, scene, frame_offset, moving_mobjects):                   │
│    384         def update_frame():                                                   │
│ File "/home/devneal/github/manim-2/manim/scene/scene.py", line 864, in play_internal │
│    861         for t in self.time_progression:                                       │
│    862             self.update_to_time(t)                                            │
│    863             if not skip_rendering:                                            │
│  ❱ 864                 self.renderer.render(self, t, self.moving_mobjects)           │
│    865             if self.stop_condition is not None and self.stop_condition():     │
│    866                 self.time_progression.close()                                 │
│    867                 break                                                         │
│ File "/home/devneal/github/manim-2/manim/renderer/opengl_renderer.py", line 394, in  │
│render                                                                                │
│    391         update_frame()                                                        │
│    392                                                                               │
│    393         if config["write_to_movie"]:                                          │
│  ❱ 394             self.file_writer.write_frame(self)                                │
│    395                                                                               │
│    396         if self.window is not None:                                           │
│    397             self.window.swap_buffers()                                        │
│ File "/home/devneal/github/manim-2/manim/scene/scene_file_writer.py", line 279, in   │
│write_frame                                                                           │
│    276         """                                                                   │
│    277         if config["use_opengl_renderer"]:                                     │
│    278             renderer = frame_or_renderer                                      │
│  ❱ 279             self.writing_process.stdin.write(                                 │
│    280                 renderer.get_raw_frame_buffer_object_data()                   │
│    281             )                                                                 │
│    282         else:                                                                 │
╰──────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'SceneFileWriter' object has no attribute 'writing_process'
```
Should this have skipped normally?

**Comment by TRoboto** [comment]
All your comments have been addressed, can you check now?

File: manim/renderer/opengl_renderer.py (hunk 1 above)
Comment (TRoboto) [reply manim/renderer/opengl_renderer.py]: I deleted it from OpenGL renderer and everything still works as expected. Please note that the decorator still exists in the Cairo renderer.

File: manim/renderer/opengl_renderer.py [hunk 2]
```diff
@@ -353,25 +357,28 @@ def init_scene(self, scene):
...
+    @handle_caching_play
+    @handle_play_like_call
     def play(self, scene, *args, **kwargs):
-        if len(args) == 0:
-            logger.warning("Called Scene.play with no animations")
-            return
-
-        # TODO: Handle data locking / unlocking.
```
[the remark this answers is not in the document]
Comment (TRoboto) [reply manim/renderer/opengl_renderer.py:361]: It has been re-added.

**Review by eulertour — APPROVED** [review]
Sorry for the wait, this looks good now

**Comment by behackl** [comment]
@TRoboto, there is a minor merge conflict now. Could you please resolve it? We can merge this PR then.

**Comment by TRoboto** [comment]
> @TRoboto, there is a minor merge conflict now. Could you please resolve it? We can merge this PR then.

@behackl conflicts are resolved. all is good now.

**Comment by behackl** [comment]
> > @TRoboto, there is a minor merge conflict now. Could you please resolve it? We can merge this PR then.
> 
> @behackl conflicts are resolved. all is good now.

Thank you! The code looks good to me, and I've also just tested your changes locally. Skipping seems to work properly, but I am not sure about caching: rendering the example scene multiple times still renders each animation (but it is possible that the problem is with the existing scene caching code).

Could you please clarify? (It's not a problem if it doesn't work yet, this PR can and should still be merged nonetheless.)

**Comment by TRoboto** [comment]
@behackl For some reason @eulertour disabled caching for OpenGL [here](https://github.com/ManimCommunity/manim/blob/1d9fdaf6296cd792e3018bd41600fe9239c806ed/manim/_config/utils.py#L1077) and [here](https://github.com/ManimCommunity/manim/blob/1d9fdaf6296cd792e3018bd41600fe9239c806ed/manim/_config/utils.py#L1088). I tried removing the two if statements and everything seems fine. Is it okay if I pushed the changes?

**Comment by behackl** [comment]
> @behackl For some reason @eulertour disabled caching for OpenGL [here](https://github.com/ManimCommunity/manim/blob/1d9fdaf6296cd792e3018bd41600fe9239c806ed/manim/_config/utils.py#L1077) and [here](https://github.com/ManimCommunity/manim/blob/1d9fdaf6296cd792e3018bd41600fe9239c806ed/manim/_config/utils.py#L1088). I tried removing the two if statements and everything seems fine. Is it okay if I pushed the changes?

I see. I've just tested removing the line for the opengl renderer, that seems to work fine. For the webgl renderer it would make sense to hear back from @eulertour -- so I'd suggest to just remove the switch for the opengl renderer, and leave scene caching disabled for the webgl renderer (for now).

**Comment by TRoboto** [comment]
> > @behackl For some reason @eulertour disabled caching for OpenGL [here](https://github.com/ManimCommunity/manim/blob/1d9fdaf6296cd792e3018bd41600fe9239c806ed/manim/_config/utils.py#L1077) and [here](https://github.com/ManimCommunity/manim/blob/1d9fdaf6296cd792e3018bd41600fe9239c806ed/manim/_config/utils.py#L1088). I tried removing the two if statements and everything seems fine. Is it okay if I pushed the changes?
> 
> I see. I've just tested removing the line for the opengl renderer, that seems to work fine. For the webgl renderer it would make sense to hear back from @eulertour -- so I'd suggest to just remove the switch for the opengl renderer, and leave scene caching disabled for the webgl renderer (for now).

I noticed that I pointed for the WebGL version so I edited my comment to include the one for OpenGL only. Anyway, everything should be good now. Thanks for pointing this issue out.

**Comment by behackl** [comment]
Excellent! Thank you for the lightning-fast responses and your effort; I'll merge this PR as soon as the pipeline is done. :-)

**Review by behackl — APPROVED** [review]

**Pull request merged**